### PR TITLE
Render Stream

### LINKED ISSUE #693: Get Renderer and Output together
Such an abstraction could have convenient drawing methods that `Canvas` currently has.

### PR DESCRIPTION
**Pull request by kvark**
Closes #693 

I can't modify the examples now, since cgmath-rs and noise-rs are broken at the moment, but the change should be backwards compatible, so we can do it any time later.

Edit: https://github.com/bjz/cgmath-rs/pull/211

Noise-rs is used from crates.io, where it's not updated yet. Not sure if it will be.
Also, gfx_gl doesn't seem to be building any more on latest nightly.

Anyway, I believe this PR is complete. Calling for reviewers - @csherratt @bvssvni @cmr @bjz

### CODE REVIEW

File: src/extra/stream.rs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    /// Command buffer type to constraint the `Renderer`.
+    type CommandBuffer: CommandBuffer<R>;
+    /// Constrained `Output` type.
+    type Output: Output<R>;
+
+    /// Get the output only.
+    fn get_output(&self) -> &Self::Output;
+
+    /// Access both of the stream components.
+    fn access(&mut self) -> (&mut Renderer<R, Self::CommandBuffer>, &Self::Output);
```
Comment (bvssvni) [inline src/extra/stream.rs]: `get_access_mut` might be better here, but I'm not sure. Discussion on get/set patterns for Piston is here https://github.com/PistonDevelopers/piston/issues/882

**Comment by bvssvni** [comment]
Looks good.

I do not like the name `Stream`, but I have no better suggestion for now.

File: src/extra/stream.rs (hunk 1 above)
Comment (kvark) [inline src/extra/stream.rs]: Problem is, it still returns the read-only `Output`, so adding mut to the function name may be confusing. If `mut` is still relevant, how about `get_all_mut`?

**Comment by kvark** [comment]
@bvssvni thanks for the review!

File: src/extra/stream.rs (hunk 1 above)
Comment (bvssvni) [inline src/extra/stream.rs]: You're right.

**Comment by bvssvni** [comment]
Merging.

**Comment by bvssvni** [comment]
@kvark Are you on IRC?

**Comment by kvark** [comment]
No, sorry, occupied with life at the moment.

**Pull request merged**